### ISSUE TITLE
Exception 'yii\db\Exception' with message 'Failed to read from socket.

### ISSUE THREAD

**Issue by yyxx9988** [yiisoft/yii2-redis#91]
### What steps will reproduce the problem?
`Yii::$app->redis->set('cron_album_pic_transfer', 0);`
### What's expected?
`OK`
### What do you get instead?
```
Exception 'yii\db\Exception' with message 'Failed to read from socket.
Redis command was: set cron_album_pic_transfer 0'

in /usr/local/nginx/html/vendor/yiisoft/yii2-redis/Connection.php:405

Stack trace:
#0 /usr/local/nginx/html/vendor/yiisoft/yii2-redis/Connection.php(394): yii\redis\Connection->parseResponse('set cron_album_...')
#1 /usr/local/nginx/html/vendor/yiisoft/yii2-redis/Connection.php(355): yii\redis\Connection->executeCommand('set', Array)
#2 /usr/local/nginx/html/console/controllers/CronJobController.php(104): yii\redis\Connection->__call('set', Array)
#3 [internal function]: console\controllers\CronJobController->actionAlbumPicTransfer()
#4 /usr/local/nginx/html/vendor/yiisoft/yii2/base/InlineAction.php(55): call_user_func_array(Array, Array)
#5 /usr/local/nginx/html/vendor/yiisoft/yii2/base/Controller.php(154): yii\base\InlineAction->runWithParams(Array)
#6 /usr/local/nginx/html/vendor/yiisoft/yii2/console/Controller.php(128): yii\base\Controller->runAction('album-pic-trans...', Array)
#7 /usr/local/nginx/html/vendor/yiisoft/yii2/base/Module.php(454): yii\console\Controller->runAction('album-pic-trans...', Array)
#8 /usr/local/nginx/html/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('cron-job/album-...', Array)
#9 /usr/local/nginx/html/vendor/yiisoft/yii2/console/Application.php(147): yii\console\Application->runAction('cron-job/album-...', Array)
#10 /usr/local/nginx/html/vendor/yiisoft/yii2/base/Application.php(375): yii\console\Application->handleRequest(Object(yii\console\Request))
#11 /usr/local/nginx/html/yii(39): yii\base\Application->run()
#12 {main}
```

### Additional info

| Q                | A
| ---------------- | ---
| Yii vesion       | 2.0.10
| PHP version      | 7.0.10
| Operating system | CentOS

**Comment by zzzeep** [comment]
```
    const ALBUM_PIC_TRANSFER_KEY = 'cron_album_pic_transfer';
    const ALBUM_PIC_TRANSFER_YES = 1;
    const ALBUM_PIC_TRANSFER_NO = 0;

    public function actionAlbumPicTransfer()
    {
        set_time_limit(0);

        $isRuning = Yii::$app->redis->get(static::ALBUM_PIC_TRANSFER_KEY);
        if (is_numeric($isRuning) && $isRuning == static::ALBUM_PIC_TRANSFER_YES) {
            echo "[". date('Y-m-d H:i:s') ."] Task [AlbumPicTransfer] is already runing\n";
            exit;
        } else {
            Yii::$app->redis->set(static::ALBUM_PIC_TRANSFER_KEY, static::ALBUM_PIC_TRANSFER_YES);
        }

        for ($i=0; $i<100; $i++) {
            $tmpPic = AlbumPicTmp::find()
                ->where(AlbumPicTmp::whereSA(false))
                ->limit(1)
                ->one();
            if ($tmpPic && isset($tmpPic->id)) {
                // download
                $localPath = Image::downloadAndSave('http://' . $tmpPic->bucket . '/' . $tmpPic->url);
                if ($localPath) {
				
					// some code
					
                    unlink($localPath);
                } else {
                    $tmpPic->note = 'error1';
                    $tmpPic->status = V4AlbumPicTmp::SD;
                    $tmpPic->update();
                }
            } else {
                break;
            }
        }

        echo "[". date('Y-m-d H:i:s') ."] Task [AlbumPicTransfer]'s loop is done\n";
        Yii::$app->redis->set(static::ALBUM_PIC_TRANSFER_KEY, static::ALBUM_PIC_TRANSFER_NO);
        exit;
    }
```

**Comment by zzzeep** [comment]
I think `set_time_limit(0)` is the KEY of this error

**Comment by cebe** [comment]
I am able to reproduce the problem by setting the redis server timeout to a value higher than 0.
However I am not sure how to solve it as there is no way to reliably detect a closed connection.

What could be done would be to introduce some retry count that could be applied to retry a failed command in case of socket error.

**Comment by cebe** [comment]
fixed by #106
